dlt/destinations/impl/sqlalchemy/db_api_client.py

```diff
@@ -0,0 +1,415 @@
+            raise self._make_database_exception(e) from e
+
+    if inspect.isgeneratorfunction(f):
+        return _wrap_gen  # type: ignore[return-value]
+    return _wrap  # type: ignore[return-value]
+
+
+class SqlaDbApiCursor(DBApiCursorImpl):
+    def __init__(self, curr: sa.engine.CursorResult) -> None:
+        # Sqlalchemy CursorResult is *mostly* compatible with DB-API cursor
+        self.native_cursor = curr  # type: ignore[assignment]
+        curr.columns
+
+        self.fetchall = curr.fetchall  # type: ignore[assignment]
+        self.fetchone = curr.fetchone  # type: ignore[assignment]
+        self.fetchmany = curr.fetchmany  # type: ignore[assignment]
+
+    def _get_columns(self) -> List[str]:
+        return list(self.native_cursor.keys())  # type: ignore[attr-defined]
+
+    # @property
+    # def description(self) -> Any:
+    #     # Get the underlying driver's cursor description, this is mostly used in tests
+    #     return self.native_cursor.cursor.description  # type: ignore[attr-defined]
+
+    def execute(self, query: AnyStr, *args: Any, **kwargs: Any) -> None:
+        raise NotImplementedError("execute not implemented")
+
+
+class DbApiProps:
+    # Only needed for some tests
+    paramstyle = "named"
+
+
+class SqlalchemyClient(SqlClientBase[Connection]):
+    external_engine: bool = False
+    dialect: sa.engine.interfaces.Dialect
+    dialect_name: str
+    dbapi = DbApiProps  # type: ignore[assignment]
+    migrations: Optional[MigrationMaker] = None  # lazy init as needed
+
+    def __init__(
+        self,
+        dataset_name: str,
+        staging_dataset_name: str,
+        credentials: SqlalchemyCredentials,
+        capabilities: DestinationCapabilitiesContext,
+        engine_args: Optional[Dict[str, Any]] = None,
+    ) -> None:
+        super().__init__(credentials.database, dataset_name, staging_dataset_name, capabilities)
+        self.credentials = credentials
+
+        if credentials.engine:
+            self.engine = credentials.engine
+            self.external_engine = True
+        else:
+            self.engine = sa.create_engine(
+                credentials.to_url().render_as_string(hide_password=False),
+                **(engine_args or {}),
+            )
+
+        self._current_connection: Optional[Connection] = None
+        self._current_transaction: Optional[SqlaTransactionWrapper] = None
+        self.metadata = sa.MetaData()
+        self.dialect = self.engine.dialect
+        self.dialect_name = self.dialect.name
+
+    def open_connection(self) -> Connection:
+        if self._current_connection is None:
+            self._current_connection = self.engine.connect()
+            if self.dialect_name == "sqlite":
+                self._sqlite_reattach_dataset_if_exists(self.dataset_name)
+        return self._current_connection
+
+    def close_connection(self) -> None:
+        if not self.external_engine:
+            try:
+                self.engine.dispose()
```

Comment:
you do not close the active connection here
https://docs.sqlalchemy.org/en/20/core/connections.html#sqlalchemy.engine.Engine.dispose

only connections in the pool will be closed by dispose. the checked out - not

IMO you must close the current connection

also do not dispose external engines.